### ISSUE TITLE
Personachat Dataset

### ISSUE THREAD

**Issue by christqoh** [facebookresearch/ParlAI#932]
Not sure if that is right here, but I have a question regarding the Personachat dataset:

According to the Paper 1155 'persons' had been created, each having at least 5 profile sentences. In train_none_original.txt there are numbers seven ... well ... some kind of chats with alternatives I guess. 

When I look into the train_self_original.txt I see numbers 1. to 11. where there are four 'your persona' sentences and seven as above. 
Also, I understood that when training on person:self the generative model (profilememory in my case) is conditioned on the self persona. However, these selves are different for every numbering of 1 to 11.

Could you please specify how the dataset is structured and how conditioning with self works?

**Comment by alexholdenmiller** [comment]
the format of the 'none' file is as follows:
```
1 x1<tab>y1<tab><tab>candidate1|candidate2|candidate3
2 x2<tab>y2<tab><tab>candidateA|candidateB|candidateC
...
```
The numbering system only serves to show how many utterance (pairs) are in that dialog. When the counter resets back to one, the next episode (conversation) starts.

In the case of the "self original" file with personas, it follows the same pattern but the first few lines don't have anything after the first utterance
```
1 xA
2 xB
3 xC
4 xD
5 x1<tab>y1<tab><tab>candidate1|candidate2|candidate3
...
```

Here, xA-D/E are personas. In ParlAI, lines without tabs are merged into the following lines, so you'll see the first example of this conversation (e.g. via `display_data.py -t personachat`) shown as `xA \n xB \n xC \n xD \n x1`. The numbering is not used except to determine when the next conversation starts.

**Comment by alexholdenmiller** [comment]
The profilememory model takes advantage of the "your persona: X\n" format of the personas in the text and extracts them from the text for separate treatment here:

https://github.com/facebookresearch/ParlAI/blob/5171ed0a3b555846cb5174c99bc6faace3ac1d99/projects/personachat/persona_seq2seq.py#L1575-L1612

**Comment by alexholdenmiller** [comment]
The model separately encodes the personas extracted in the function and uses the output from that encoder to do attention over the personas.

**Comment by christqoh** [comment]
Thank you @alexholdenmiller for your reply. I feel like there is a need to clear up some wording first: 
Starting from the `train_none_original_no_cands.txt`, the model learns the mapping from input sequence x to output sequence y. 

Extending this with `candidates`, where are candidate1|candidate2|candidate3... originating in? Are those simply possible ways turkeys used to continue the conversation?

Further on with `train_self_original.txt` you say that _Here, xA-D/E are personas._ What/is there a reason you added the /E as it is not shown in your example right above that? 
The first utterances being the personas, what exactly are those? According to the paper, chats are based on two virtual persons each having at least five profile sentences. Here, there are 4 utterances followed by the chat. Where are they originating in? One of the 'persons' profiles? Are those utterances even based on the same 'person' (out of the 1155)?

**Comment by christqoh** [comment]
that clears things up, thank you @alexholdenmiller. However, I am not sure I understand your third and second to last sentences. 

So in theory a arbitrary number of "your persona" utterances could be specified (a very complex person represented) as long as it sticks to the representation standards?

**Comment by alexholdenmiller** [comment]
yes I mean we certainly can't represent a human's entire personality in five sentences!
whether a model could successfully handle that many is a separate modeling problem, of course...

**Comment by alexholdenmiller** [comment]
Note that when we include datasets in ParlAI, we try as much as possible to format them in a way that a new model that has not been run on this dataset should have as small a barrier of entry as possible. Similarly, we try to format them such that a human could participate in the task as well. (These are why we use strings instead of tensors in most places!)

For example, parlai/agents/seq2seq has zero code for extracting and interpreting the personas, but because the personas were just in the text of the examples it was immediately able to train on it (and works surprisingly well, though of course I believe a separate persona encoder has the potential for higher performance!).

**Comment by christqoh** [comment]
thank you, could you still please specify your reply (which way around is it?):
> where the "your persona" messages are the persona of <person 1> and "hello, how are you doing tonight ?" is being said to <person 1> by , say.
during the data collection, was provided the sentences shown via "your persona", and then typed the first message ("hello, how are you doing tonight?").

Further on, how is self and other defined in the dataset? Is self simply the one starting the conversation?

**Comment by alexholdenmiller** [comment]
<img width="513" alt="screenshot 2018-06-29 15 13 13" src="https://user-images.githubusercontent.com/1141354/42110405-f735373c-7bae-11e8-8669-482e19815708.png">
(image from the paper)
first, we collected Persona1 and Persona2 as standalone personalities.

then, we paired two humans, who were shown their corresponding persona.
then the conversation starts.

in the data, let's say the model is in the position of Person 2.
it will receive: "your persona: I eat sardines for breakfast daily \n your persona: I just had surgery... \n Hi I am Dr. Jenson. What is your name?" and the model should learn to respond "Miss Dawes."

**Comment by christqoh** [comment]
> it will receive: "your persona: I eat sardines for breakfast daily \n your persona: I just had surgery... \n Hi I am Dr. Jenson. What is your name?" and the model should learn to respond "Miss Dawes."

Thanks @alexholdenmiller. So when I use the `self` training data, my model learns to be Person 1 of the conversations? And vice versa for `other`?

**Comment by alexholdenmiller** [comment]
No, in the "personachat" task we only show the data from the perspective of Person 2. for the "other" task, it's also from the perspective of person 2, but you see the person1's persona instead of your own. (although of course if you aren't using the parlai training loop and are looking directly at the files you can learn it from either perspective.)

in the convai2 task, we apply a little bit nicer preprocessing, add more candidates for ranking, and show the dataset from both perspectives. in the example above, we will also have an example where Persona1 will be shown to the model followed by "__SILENCE__" (and the agent should learn to say "Hi I am Dr. Jenson. What is your name?")

**Comment by christqoh** [comment]
great, thank you so much Alex!

**Comment by Hilal-Urun** [comment]
Hi, I want to ask that can we manage the personas like there are around 1182 personas . For example can we group or when we are chatting can we select the bot's persona?

**Comment by OneDirection9** [comment]
Hi all,

The discussions above have a lot of help for me to understand the format of the dataset file. And I have a question, take following as an example:

```
1. your persona: A
2. your persona: B
3. your persona: C
4. your persona: D
5. partner's persona: X
6. partner's persona: Y
7. partner's persona: Z
8. ...
9. x1\ty1
10. x2\ty2
11. x3\ty3
```

Does the x1,x2,x3 and so on are say by Person 1 ( described by the `your persona` fields). The y1,y2,y3, and so on are for partner's persona?

Thanks!

**Comment by weownthenight** [comment]
> Hi all,
> 
> The discussions above have a lot of help for me to understand the format of the dataset file. And I have a question, take following as an example:
> 
> ```
> 1. your persona: A
> 2. your persona: B
> 3. your persona: C
> 4. your persona: D
> 5. partner's persona: X
> 6. partner's persona: Y
> 7. partner's persona: Z
> 8. ...
> 9. x1\ty1
> 10. x2\ty2
> 11. x3\ty3
> ```
> 
> Does the x1,x2,x3 and so on are say by Person 1 ( described by the `your persona` fields). The y1,y2,y3, and so on are for partner's persona?
> 
> Thanks!

I checked `test_both_original.txt` and found that x1, x2, x3 are said by partner's persona. Like this:

```
1 your persona: i drive a ford pickup truck.
2 your persona: i am very conservative.
3 your persona: my family lives down the street from me.
4 your persona: i go to church every sunday.
5 your persona: i have three guns and love hunting.
6 partner's persona: i am an old man.
7 partner's persona: i have a wheelchair that is modded to go very fast over many terrains.
8 partner's persona: i used to be in the military.
9 partner's persona: i only have one leg.
10 how are you ? being an old man , i am slowing down these days	hi , my dad is old as well , they live close to me and i see them often
```
